### PR TITLE
1.0.0-beta candidate - etsy v3 api + oAuth helper

### LINKED ISSUE #36: findAllShopSections : listingId argument not used
the `listingId` argument is not required here and need to be removed

https://github.com/creharmony/node-etsy-client/blob/107f3d2a9ab20ae1dd31e50356ddf02fe8d65b4a/src/EtsyClient.js#L60

NB/ that's a breaking change

### LINKED ISSUE #39: Improvement: Update Etsy API to v3
Since etsys API v2 will be discontinued and stop working by the end of the year as stated in the etsy documentation ( https://www.etsy.com/developers/documentation/getting_started/edd#section_update_your_app ), it would be wise to update the client to use etsys new v3 API - which is similar but not exactly the same.

Details can be found here:
https://developers.etsy.com/documentation/

Great job here btw! It's nice to see an updated etsy node client.

### PR DESCRIPTION
**Pull request by creharmony**
Fix #39 Update Etsy API to v3 - special thanks to [icemansparks](https://github.com/icemansparks) 🙏 
    
- migrate existing v2 functions to v3 apis (cf some remarks on #39)
- add very first manual oauth2 tests
- add manual steps + util function to get oauth2 access token

Fix #36 v2 findAllShopSections : listingId argument not used

---

All reviewers are welcome : without feedbacks, this PR may be merged 3 april but next version will be beta and incomming changes / issues / PR are always welcome !

- the typescript part still suffers from a lack of testing - @geopic 

all feedback (positive/negative) are welcome

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,10 +40,10 @@ npm install node-etsy-client
 
 then let's go, here is a `sample.js`:
 ```
-const EtsyClient = require('node-etsy-client')
+const EtsyClientV2 = require('node-etsy-client')
```
Comment (creharmony) [inline README.md:43]: v3

File: README.md [hunk 2]
```diff
@@ -35,10 +40,10 @@ npm install node-etsy-client
...
 async function doIt() {
-  var client = new EtsyClient();
-  var shops = await client.getShops({'limit':10});
+  var client = new EtsyClientV2();
```
Comment (creharmony) [inline README.md:45]: v3

File: README.md [hunk 3]
```diff
@@ -47,21 +52,22 @@ You could play mocha tests to get more examples (cf. next section).
...
-var client = new EtsyClient({apiKey:'mSecretHere'});
+var client = new EtsyClientV3({apiKey:'mSecretHere'});
 ```
 
 ## Advanced usage
 
 
 ### Etsy client options
-This section describes EtsyClient available options.
+This section describes EtsyClientV2 available options.
```
Comment (creharmony) [inline README.md:62]: v2 or v3

File: README.md [hunk 4]
```diff
@@ -86,9 +92,42 @@ To apply rate limit of 10 query per seconds (with wait on unavailable slot),
 add `etsyRateMaxQueries` option:
 
 ```
-var client = new EtsyClient({apiKey:'mSecretHere', etsyRateMaxQueries:10});
+var client = new EtsyClientV2({apiKey:'mSecretHere', etsyRateMaxQueries:10});
```
Comment (creharmony) [inline README.md:95]: promote v3

File: README.md [hunk 5]
```diff
@@ -6,16 +6,21 @@ NodeJs Etsy [ReST API](https://www.etsy.com/developers/documentation) Client.
...
 - findAllShops
 - getShop
 - findAllShopSections
-- findAllShopListingsActive 
+- findAllShopListingsActive
 - getListing
 - getVariationImages
 - findAllListingImages
-- getInventory 
+- getInventory
```
Comment (creharmony) [inline README.md:23]: verify list

File: src/EtsyClientV3.d.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
+declare module 'node-etsy-client' {
+  type EtsyClientOptions = Partial<{
+    apiUrl: string;
+    apiKey: string;
+    lang: string;
+    shop: string;
+    shopId: number;
+  }>
+
+  export default class EtsyClient {
```
Comment (creharmony) [inline src/EtsyClientV3.d.ts:10]: name v3 suffix

File: src/EtsyClientV3.js [hunk 7]
```diff
@@ -0,0 +1,255 @@
+import queryString from 'query-string';
+import fetch from 'node-fetch';
+import SusiRali from 'susi-rali';
+
+const DEFAULT_DRY_MODE = false;// DEBUG // if true, then print fetch onto console instead of calling etsy
+
+/**
+ * this utility class implement ETSY (some) methods documented here:
+ * https://www.etsy.com/developers/documentation > API Reference
```
Comment (creharmony) [inline src/EtsyClientV3.js:9]: check v3 link

File: src/EtsyClientV3.js [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+const DEFAULT_DRY_MODE = false;// DEBUG // if true, then print fetch onto console instead of calling etsy
+
+/**
+ * this utility class implement ETSY (some) methods documented here:
+ * https://www.etsy.com/developers/documentation > API Reference
+ **/
+class EtsyClientV3 {
+
+  static debug = process.env.ETSY_DEBUG && process.env.ETSY_DEBUG === "true";
```
Comment (creharmony) [inline src/EtsyClientV3.js:13]: talk about debug mode in readme

File: src/EtsyClientV3.js [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+    this._assumeApiKey();
+    this.shopId = "shopId" in options ? options.shopId : process.env.ETSY_SHOP_ID;
+    this.accessToken = "accessToken" in options ? options.accessToken : null;
+    this.lang   = "lang" in options ? options.lang : process.env.ETSY_LANG;
+    // configure rate limit on etsy call : max <etsyRateMaxQueries> per <etsyRateWindowSizeMs> ms
+    // Etsy rate limit doc (10/sec) : https://www.etsy.com/developers/documentation/getting_started/api_basics#section_rate_limiting
+    this.etsyRateWindowSizeMs = "etsyRateWindowSizeMs" in options ? options.etsyRateWindowSizeMs : (process.env.ETSY_RATE_WINDOWS_SIZE_MS  || 1000);
+    this.etsyRateMaxQueries   = "etsyRateMaxQueries" in options ? options.etsyRateMaxQueries : (process.env.ETSY_RATE_MAX_QUERIES || null);
+    this.dryMode              = "dryMode" in options ? options.dryMode : ("true" === process.env.ETSY_DRY_MODE || DEFAULT_DRY_MODE);
+    this.initRateLimiter();
```
Comment (creharmony) [inline src/EtsyClientV3.js:31]: test vith oauth token and without rate limit

**Review by creharmony — COMMENTED** [review]
review to continue

File: package.json [hunk 10]
```diff
@@ -6,10 +6,14 @@
...
   "scripts": {
-    "test": "mocha tests/*.test.js",
-    "longTest": "set ETSY_LONG_TEST=1&& mocha --timeout 120000 tests/unauthenticated_rate.test.js",
+    "testV2": "mocha tests/v2/*.test.js",
+    "longTestV2": "set ETSY_LONG_TEST=1&& mocha --timeout 120000 tests/v2/unauthenticated_rate.test.js",
+    "testV3": "mocha tests/v3/*.test.js",
+    "longTestV3": "set ETSY_LONG_TEST=1&& mocha --timeout 120000 tests/v3/unauthenticated_rate.test.js",
+    "test": "mocha tests/**/*.test.js",
+    "manual": "mocha tests/**/*.tst.js",
     "cover": "nyc --reporter text --reporter cobertura --reporter html --reporter=lcov --lines 66 mocha --exit --unhandled-rejections=strict tests/*.test.js",
```
Comment (creharmony) [inline package.json:15]:
```suggestion
    "cover": "nyc --reporter text --reporter cobertura --reporter html --reporter=lcov --lines 66 mocha --exit --unhandled-rejections=strict tests/**/*.test.js",
```

File: src/EtsyClientV2.js [hunk 11]
```diff
@@ -8,7 +8,7 @@ const DEFAULT_DRY_MODE = false;// DEBUG // if true, then print fetch onto consol
  * this utility class implement ETSY (some) methods documented here:
  * https://www.etsy.com/developers/documentation > API Reference
```
Comment (creharmony) [inline src/EtsyClientV2.js:9]: add V2

try to find V2 link

**Comment by boly38** [comment]
would be great to fix nyc too
- #42

**Pull request merged**